Support showing opcodes in the disassembly

lebr0nli opened the pull request:
This PR is trying to implement the feature that #1422 mentioned.
By setting `nearpc-num-opcode-bytes`(default is 0) and `nearpc-opcode-seperator-bytes` (default is 1):
- `set nearpc-num-opcode-bytes 9`
<img width="1217" alt="截圖 2023-01-24 下午8 25 49" src="https://user-images.githubusercontent.com/61896187/214291494-023e86cb-d718-4aeb-ab15-8533acff4d19.png">

- `set nearpc-num-opcode-bytes 4`

<img width="1036" alt="截圖 2023-01-24 下午8 26 34" src="https://user-images.githubusercontent.com/61896187/214291517-2ffc415f-8f88-4be9-8c04-cc2f70cddad0.png">

- `set nearpc-num-opcode-bytes 9` + `set nearpc-opcode-seperator-bytes 0`
 <img width="1119" alt="截圖 2023-01-24 下午8 36 51" src="https://user-images.githubusercontent.com/61896187/214293493-3e27d77a-7b58-4067-82e1-5aad8c209654.png">

lebr0nli committed:
Add an option to show opcodes in the disassembly

gsingh93 commented on pwndbg/gdblib/nearpc.py:
```diff
@@ -162,7 +165,19 @@ def nearpc(pc=None, lines=None, emulate=False, repeat=False) -> List[str]:
+            opcodes = opcodes.ljust(align, " ")
+            if pwndbg.gdblib.config.highlight_pc and instr.address == pwndbg.gdblib.regs.pc:
+                opcodes = C.highlight(opcodes)
+            line = " ".join((prefix, address_str, opcodes, symbol, asm))
```
nit: Instead of duplicating this line here and in the `else` clause, we can probably add a `opcodes = ""` above the if statement and then move this right after the if statement, so both code paths use the same code to construct the output.

lebr0nli replied:
Sure, I'll remove this `else` clause.

gsingh93 commented on pwndbg/gdblib/nearpc.py:
```diff
@@ -162,7 +165,19 @@ def nearpc(pc=None, lines=None, emulate=False, repeat=False) -> List[str]:
+        if show_opcodes_size > 0:
+            opcodes = instr.bytes.hex()
+            opcodes = opcodes[: show_opcodes_size * 2]
+            align = show_opcodes_size * 2 + 10
+            if len(instr.bytes) > show_opcodes_size:
+                opcodes += pwndbg.color.gray("...")
+                align += 9  # len(pwndbg.color.gray(""))
```
Can you add some comments explaining how the alignment is calculated? So if we show 9 opcode bytes (18 characters) we'll pad the end of it by 10 spaces, but if we add a '...' we pad the end by 19 spaces?

lebr0nli replied:
> Can you add some comments explaining how the alignment is calculated?

Sure.

> So if we show 9 opcode bytes (18 characters) we'll pad the end of it by 10 spaces, but if we add a '...' we pad the end by 19 spaces?

This 9 byte is because the gray color uses 9 invisible characters, so we need to extend the alignment for it to make it look the same.

gsingh93 commented on pwndbg/gdblib/nearpc.py:
```diff
@@ -54,7 +54,10 @@ def ljust_padding(lst):
+    "nearpc-show-args", True, "whether to show call arguments below instruction"
+)
+show_opcodes_size = pwndbg.gdblib.config.add_param(
+    "nearpc-opcodes-size", 0, "number of bytes of opcodes to print for each instruction"
```
nit: "number of opcode bytes to print for each instruction"

gsingh93 replied:
And maybe the option name should be `nearpc-opcode-bytes`? Not sure.

disconnect3d replied:
or: `nearpc-show-opcode-bytes`?

lebr0nli replied:
Ok, I'll use `nearpc-show-opcode-bytes`

gsingh93 replied:
`nearpc-show-opcode-bytes` makes it sound like a boolean. `nearpc-num-opcode-bytes` is another alternative.

lebr0nli replied:
> `nearpc-show-opcode-bytes` makes it sound like a boolean. `nearpc-num-opcode-bytes` is another alternative.

Ok, I used `nearpc-num-opcode-bytes` now.

gsingh93 reviewed (commented):
This is great, thanks. Could you update the `nearpc` test to also test when `nearpc-opcode-size` is non-zero?

This doesn't need to be done in this PR, but I think having an option for adding spaces between each opcode byte would be nice.

lebr0nli committed:
Refactor the code for showing opcodes

Remove `else` caluse (https://github.com/pwndbg/pwndbg/pull/1537/files#r1083835300)

lebr0nli committed:
Add comment about the alignment for "..." when showing opcodes

lebr0nli committed:
Update the config for opcodes

Rename `nearpc-opcodes-size` to `nearpc-opcode-bytes` and update the description

lebr0nli committed:
Sync the highlighting behaviour

Only highlight the opcodes when `prefix`, `address_str` and `symbol` have been highlighted

lebr0nli committed:
Add a config to set the number of spaces between opcode bytes

lebr0nli committed:
Avoid extra space if opcodes is empty

lebr0nli committed:
Avoid negative opcode config

lebr0nli committed:
Add tests for opcode

lebr0nli commented:
> Could you update the nearpc test to also test when nearpc-opcode-size is non-zero?

Sure.

> I think having an option for adding spaces between each opcode byte would be nice

Yeah, after some testing, seems like the readability with a space is better.
I make it default to one space for now and use `nearpc-opcode-seperator-bytes` for controlling it, but open to other suggestions.

lebr0nli committed:
Bug fix for `gdb.PARAM_ZUINTEGER*` with GDB < 9

GDB < 9 does not support PARAM_ZUINTEGER*, so we implement it by ourselves for consistency

lebr0nli committed:
Rename config

Replace `nearpc-opcode-bytes` with `nearpc-num-opcode-bytes`

disconnect3d commented on pwndbg/gdblib/config.py:
```diff
@@ -66,6 +79,15 @@ def get_set_string(self) -> str:
         else:
+            if self._custom_param_class:
+                if (self._custom_param_class == "PARAM_ZUINTEGER" and self.value < 0) or (  # type: ignore
+                    self._custom_param_class == "PARAM_ZUINTEGER_UNLIMITED" and self.value < -1  # type: ignore
```
Since this is making more and more complex and less and less clear, shall we maybe do something along the lines of:

```py
def __get_set_string_gdb_le_9(self): ...
def __get_set_string_gdb_gte_9(self): ...

if gdb_version >= 9:
    get_set_string = __get_set_string_gdb_gte_9
else:
    get_set_string = __get_set_string_gdb_le_9
```
@gsingh93 WDYT?

gsingh93 replied:
+1, I think that'll be less complicated and cleaner.

lebr0nli committed:
Refactor `pwndbg.gdblib.config.Parameter`'s methods

Now we initialize the different methods at the start for GDB < 9 and GDB >=9, to avoid running if-else statements every time and make it looks clearer. (https://github.com/pwndbg/pwndbg/pull/1537#discussion_r1086215844)

disconnect3d reviewed (approved):
@gsingh93 I think we can merge this now?

gsingh93 commented:
Yup, lgtm.
